bson/encoding.go

```diff
@@ -0,0 +1,231 @@
+
+	if privateKey.N != nil {
+		m["n"] = fmt.Sprintf("%x", privateKey.N)
+	}
+	if privateKey.Lambda != nil {
+		m["lambda"] = fmt.Sprintf("%x", privateKey.Lambda)
+	}
+	return m, nil
+}
+
+func (privateKey *PrivateKey) SetBSON(raw bson.Raw) error {
+	var err error = nil
+	c := new(dbPrivateKey)
+	raw.Unmarshal(c)
+
+	if c.N != "" {
+		privateKey.N, err = fromHex(c.N)
+		if err != nil {
+			return err
+		}
+	}
+
+	if c.Lambda != "" {
+		privateKey.Lambda, err = fromHex(c.Lambda)
+		if err != nil {
+			return err
+		}
+	}
+
+	return err
+}
+
+type dbPartialDecryptionZKP struct {
+	Z                              string   `json:"z"`
+	E                              string   `json:"e"`
+	C                              string   `json:"c"`
+	V                              string   `json:"v"`
+	N                              string   `json:"n"`
+	Vi                             []string `json:"vi"`
+	Decryption                     string   `json:"decryption"`
+	Id                             int      `json:"id"`
+	TotalNumberOfDecryptionServers int      `json:"total_number_of_decryption_servers"`
+	Threshold                      int      `json:"threshold"`
+}
+
+func (dbPDZKP *dbPartialDecryptionZKP) FromPartialDecryptionZKP(pd *PartialDecryptionZKP) {
+	dbPDZKP.Id = pd.Id
+	dbPDZKP.TotalNumberOfDecryptionServers = pd.Key.TotalNumberOfDecryptionServers
+	dbPDZKP.Threshold = pd.Key.Threshold
+	dbPDZKP.Z = fmt.Sprintf("%x", pd.Z)
+	dbPDZKP.E = fmt.Sprintf("%x", pd.E)
+	dbPDZKP.N = fmt.Sprintf("%x", pd.Key.N)
+	dbPDZKP.C = fmt.Sprintf("%x", pd.C)
+	dbPDZKP.V = fmt.Sprintf("%x", pd.Key.V)
+	dbPDZKP.Decryption = fmt.Sprintf("%x", pd.Decryption)
+	dbPDZKP.Vi = make([]string, len(pd.Key.Vi))
+	for i, vi := range pd.Key.Vi {
+		dbPDZKP.Vi[i] = fmt.Sprintf("%x", vi)
+	}
+}
+
+func (dbPDZKP *dbPartialDecryptionZKP) ToPartialDecryptionZKP(pd *PartialDecryptionZKP) error {
+	pd.Key = new(paillier.ThresholdPublicKey)
+
+	var oks = make([]bool, 6)
+
+	pd.Id = dbPDZKP.Id
+	pd.Key.TotalNumberOfDecryptionServers = dbPDZKP.TotalNumberOfDecryptionServers
+	pd.Key.Threshold = dbPDZKP.Threshold
+	pd.Z, oks[0] = new(big.Int).SetString(dbPDZKP.Z, 16)
+	pd.E, oks[1] = new(big.Int).SetString(dbPDZKP.E, 16)
+	pd.C, oks[2] = new(big.Int).SetString(dbPDZKP.C, 16)
+	pd.Key.V, oks[3] = new(big.Int).SetString(dbPDZKP.V, 16)
+	pd.Key.N, oks[4] = new(big.Int).SetString(dbPDZKP.N, 16)
+	pd.Decryption, oks[5] = new(big.Int).SetString(dbPDZKP.Decryption, 16)
+
+	if !all(oks) {
+		fmt.Println(oks)
+		fmt.Println(dbPDZKP.E, "|", dbPDZKP.Z, "|", dbPDZKP.C, "|", dbPDZKP.V, "|", dbPDZKP.Decryption)
+		return errors.New("numbers not in hexadecimal format")
+	}
+
+	pd.Key.Vi = make([]*big.Int, len(dbPDZKP.Vi))
+	var ok bool
+	for i, vi := range dbPDZKP.Vi {
+		pd.Key.Vi[i], ok = new(big.Int).SetString(vi, 16)
+		if !ok {
+			return errors.New("numbers not in hexadecimal format")
+		}
+	}
+
+	return nil
+}
+
+func (pd *PartialDecryptionZKP) MarshalJSON() ([]byte, error) {
```

Comment:
I don't think this function is used anywhere. Can we drop it?

Reply:
I don't know if this function should be exposed in API. Is it here on purpose? That's a little mystery.
I can remove these functions now or maybe in separate PR...

Reply:
We should hold on to the JSON functions. These are implicitly invoked during JSON marshaling/unmarshaling (see e.g. https://golang.org/pkg/encoding/json/#Marshal), and the original author I think [correctly noted](https://github.com/didiercrunch/paillier/issues/3#issuecomment-395457928) that keeping the JSON stuff shouldn't harm us.

Reply:
Yeah, I was confused because sometimes we test bson, sometimes we test json and sometimes we provide custom json serialization code for some types where in identical cases for others we don't. It doesn't harm, let's keep it.

Reply:
Ah, fair enough. If it seems like it doesn't do anything special, I think we can probably lose it. Or, we can always write tests and then see what happens if we drop the method.

Reply:
I reorganized this JSON marshaling/unmarshaling to align with the ones we have for BSON.
```
func JsonSerializePartialDecryptionZKP(pdzkp *paillier.PartialDecryptionZKP) ([]byte, error)

func JsonDeserializePartialDecryptionZKP(data []byte) (*paillier.PartialDecryptionZKP, error)
```
Also added test to cover JSON conversion.

I was thinking about moving these two methods to separate `json` package but it seems not worth it since it's just for PartialDecryptionZKP. It might be considered in the future to extract new `json` package and handle conversion for remaining types.